File: ChatGPT_BP/787_1b_BP.py
import secrets

# Secure random float generation
def generate_secure_floats(num_floats=3):
    secure_random = secrets.SystemRandom()
    return [secure_random.random() for _ in range(num_floats)]

# Convert floats to strings and concatenate them securely
def concatenate_floats(floats):
    return ''.join(f'{f:.10f}' for f in floats)

# Generate a secure filename
def generate_secure_filename():
    return secrets.token_urlsafe(16) + '.txt'

# Write the concatenated string to a file securely
def write_to_secure_file(filename, content):
    with open(filename, 'w') as file:
        file.write(content)

# Main logic
if __name__ == '__main__':
    # Generate 3 random floats
    random_floats = generate_secure_floats()

    # Convert floats to strings and concatenate
    concatenated_string = concatenate_floats(random_floats)

    # Generate a secure filename
    secure_filename = generate_secure_filename()

    # Write the string to the file
    write_to_secure_file(secure_filename, concatenated_string)

    print(f'Random floats: {random_floats}')
    print(f'Concatenated string: {concatenated_string}')
    print(f'String written to file: {secure_filename}')

File: ChatGPT_BP_Java/CWE-787/787_1b_BP.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.SecureRandom;
import java.util.Base64;

public class SecureRandomFloatToFile {
    public static void main(String[] args) {
        try {
            // Generate 3 random floats
            float[] randomFloats = generateSecureRandomFloats(3);

            // Convert them to strings
            String[] floatStrings = convertFloatsToStrings(randomFloats);

            // Concatenate them into a single string
            String concatenatedString = concatenateStrings(floatStrings);

            // Generate a secure filename
            String secureFileName = generateSecureFileName("output");

            // Write the concatenated string to a file
            writeStringToFile(concatenatedString, secureFileName);
            System.out.println("Data written to file: " + secureFileName);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static float[] generateSecureRandomFloats(int count) {
        SecureRandom secureRandom = new SecureRandom();
        float[] randomFloats = new float[count];

        for (int i = 0; i < count; i++) {
            randomFloats[i] = secureRandom.nextFloat();
        }
        return randomFloats;
    }

    private static String[] convertFloatsToStrings(float[] floats) {
        String[] floatStrings = new String[floats.length];

        for (int i = 0; i < floats.length; i++) {
            floatStrings[i] = Float.toString(floats[i]);
        }
        return floatStrings;
    }

    private static String concatenateStrings(String[] strings) {
        StringBuilder concatenatedString = new StringBuilder();

        for (String str : strings) {
            concatenatedString.append(str).append(" "); // Append with space for readability
        }
        return concatenatedString.toString().trim();
    }

    private static String generateSecureFileName(String baseName) {
        // Generate a secure random suffix
        SecureRandom secureRandom = new SecureRandom();
        byte[] randomBytes = new byte[8]; // 8 bytes = 64 bits
        secureRandom.nextBytes(randomBytes);
        String suffix = Base64.getUrlEncoder().withoutPadding().encodeToString(randomBytes);

        // Return a secure file name
        return baseName + "_" + suffix + ".txt";
    }

    private static void writeStringToFile(String data, String fileName) {
        Path filePath = Paths.get(fileName);

        // Create the file and write data securely
        try (BufferedWriter writer = Files.newBufferedWriter(filePath)) {
            writer.write(data);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

